vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx

```diff
@@ -0,0 +1,142 @@
+  private _restoring: boolean;
+  private _timeout?: number;
+
+  static ID = 'trace-explorer-time-range-data-widget';
+  static LABEL = 'Time Range Data';
+
+  constructor(props: TimeRangeDataWidgetProps) {
+      super(props);
+      this._signalHandler = new VsCodeMessageManager();
+      this._reactRef = React.createRef();
+      this._restoring = false;
+
+      window.addEventListener('message', event => {
+
+          const { command, data } = event.data;
+
+          switch (command) {
+          case VSCODE_MESSAGES.RESTORE_VIEW:
+              this._restoring = true;
+              this.restoreState();
+              return;
+          case VSCODE_MESSAGES.RESTORE_COMPLETE:
+              this._restoring = false;
+              break;
+          case VSCODE_MESSAGES.TRACE_VIEWER_TAB_CLOSED:
+              signalManager().fireCloseTraceViewerTabSignal(data);
+              break;
+          case VSCODE_MESSAGES.EXPERIMENT_SELECTED:
+              signalManager().fireExperimentSelectedSignal(
+                data?.wrapper ? convertSignalExperiment(JSONBig.parse(data.wrapper)) : undefined
+              );
+              break;
+          case VSCODE_MESSAGES.EXPERIMENT_UPDATED:
+              signalManager().fireExperimentUpdatedSignal(
+                  convertSignalExperiment(JSONBig.parse(data.wrapper))
+              );
+              break;
+          case VSCODE_MESSAGES.EXPERIMENT_CLOSED:
+              signalManager().fireExperimentClosedSignal(
+                  convertSignalExperiment(JSONBig.parse(data.wrapper))
+              );
+              break;
+          case VSCODE_MESSAGES.SELECTION_RANGE_UPDATED:
+              signalManager().fireSelectionRangeUpdated(JSONBig.parse(data));
+              break;
+          case VSCODE_MESSAGES.VIEW_RANGE_UPDATED:
+              signalManager().fireViewRangeUpdated(JSONBig.parse(data));
+              break;
+          }
+          this.storeState();
+      });
+  }
+
+  componentWillUnmount = (): void => {
+      signalManager().off(Signals.REQUEST_SELECTION_RANGE_CHANGE, this.onRequestSelectionChange);
+  };
+
+  onRequestSelectionChange = (payload: TimeRangeUpdatePayload): void => {
+      this._signalHandler.requestSelectionRangeChange(payload);
+  };
+
+  componentDidMount = (): void => {
+      this._signalHandler.notifyReady();
+      signalManager().on(Signals.REQUEST_SELECTION_RANGE_CHANGE, this.onRequestSelectionChange);
+
+  };
+
+  storeState = (): void => {
+      if (!this._reactRef.current || this._restoring) {
+          return;
+      }
+      if (this._reactRef.current.isBusy) {
+          if (this._timeout) {
+              window.clearTimeout(this._timeout);
+          }
+          this._timeout = window.setTimeout(this.storeState, 100);
+          return;
+      }
+      const { activeData, experimentDataMap } = this._reactRef.current.persistedState;
+      vscode.setState({
+          activeData: JSONBig.stringify(activeData),
```

Comment:
So, if you serialize using JSONBig, all bigint variables are serialized in a special way so that when calling JSONBig.parse() a bigint is recreated. However, if the bigint value fits into a javascript number when calling JSONBig.stringify() it won't serialize it that special way and when calling JSON.big.parse() the corresponding variable is NOT a bigint. This can lead to problems later on.

So, this is the reason why the tsp-typescript-client has special code to account for that and the vscode-trace-extension has some special code for serializing/deserializing Experiment. See vscode-signal-converter.ts.

So, you will need take care of this when doing JSONBig.parse(),

BTW, we plan to streamline both vscode-signal-converter.ts and tsp-typescript-client for the bigint handling (not scope of this PR).